src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServer.cs

```diff
@@ -213,28 +215,41 @@ public void Start<TContext>(IHttpApplication<TContext> application)
             }
         }
 
-        public void Dispose()
+        // Graceful shutdown if possible
+        public async Task StopAsync(CancellationToken cancellationToken)
         {
+            _isStopping = true;
```

Comment:
If you're going to have an `_isStopping` field, you might as well guard against parallel stops. Might want something like `if (Interlocked.Exchange(ref hasStartedStopping, 1) == 0) ...`.